controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -0,0 +1,241 @@
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha3"
+
+	pkgcfg "github.com/vmware-tanzu/vm-operator/pkg/config"
+	pkgctx "github.com/vmware-tanzu/vm-operator/pkg/context"
+	"github.com/vmware-tanzu/vm-operator/pkg/patch"
+	"github.com/vmware-tanzu/vm-operator/pkg/prober"
+	"github.com/vmware-tanzu/vm-operator/pkg/record"
+)
+
+const (
+	finalizerName = "virtualmachinereplicaset.vmoperator.vmware.com"
+)
+
+// AddToManager adds this package's controller to the provided manager.
+func AddToManager(ctx *pkgctx.ControllerManagerContext, mgr manager.Manager) error {
+	var (
+		controlledType     = &vmopv1.VirtualMachineReplicaSet{}
+		controlledTypeName = reflect.TypeOf(controlledType).Elem().Name()
+
+		controllerNameShort = fmt.Sprintf("%s-controller", strings.ToLower(controlledTypeName))
+		controllerNameLong  = fmt.Sprintf("%s/%s/%s", ctx.Namespace, ctx.Name, controllerNameShort)
+	)
+
+	r := NewReconciler(
+		ctx,
+		mgr.GetClient(),
+		ctrl.Log.WithName("controllers").WithName(controlledTypeName),
+		record.New(mgr.GetEventRecorderFor(controllerNameLong)))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(controlledType).
+		Owns(&vmopv1.VirtualMachine{}).
+		Watches(&vmopv1.VirtualMachine{},
+			handler.EnqueueRequestsFromMapFunc(r.VMToReplicaSets(ctx)),
+		).
+		// TODO: Should we also watch the volumes so we get an event for changes
+		// to a volume mounted to a VM that is part of a ReplicaSet?
+		WithOptions(controller.Options{MaxConcurrentReconciles: ctx.MaxConcurrentReconciles}).
+		Complete(r)
+
+}
+
+// VMToReplicaSets is a mapper function to be used to enqueue requests for reconciliation for VirtualMachineSetReplicaSets that might adopt a VM.
+func (r *Reconciler) VMToReplicaSets(ctx *pkgctx.ControllerManagerContext) func(_ context.Context, o client.Object) []reconcile.Request {
+
+	// If the replica VM already has a controller reference, return early. If
+	// not, enqueue a reconcile if the VM matches the selector specified in the
+	// ReplicaSet. Note that we don't support adopting a replica since that
+	// might involve relocating it across compute and storage entities -- which
+	// is not supported. Instead, we emit a Condition on the ReplicaSet resource
+	// indicating it has orphaned replicas.
+	return func(_ context.Context, o client.Object) []reconcile.Request {
+		result := []ctrl.Request{}
+		vm, ok := o.(*vmopv1.VirtualMachine)
+		if !ok {
+			panic(fmt.Sprintf("Expected a VirtualMachine, but got a %T", o))
+		}
+
+		// If this VM already has a controller reference, no reconciles are
+		// queued.
+		for _, ref := range vm.ObjectMeta.GetOwnerReferences() {
+			if ref.Controller != nil && *ref.Controller {
+				return nil
+			}
+		}
+
+		vmrsss, err := r.getReplicaSetsForVM(ctx, vm)
+		if err != nil {
+			ctx.Logger.Error(err, "Failed getting VM ReplicaSets for VM")
+			return nil
+		}
+
+		if len(vmrsss) == 0 {
```

Comment:
Could get rid of this block. If return a nil vs an empty slice matters, change result into `var result []ctrl.Request`